skimage/feature/fisher_vector.py

```diff
@@ -0,0 +1,223 @@
+        Keyword arguments that can be passed into the underlying scikit-learn
+        GaussianMixture class.
+
+    Returns
+    -------
+    gmm : sklearn.mixture.GaussianMixture
+        The estimated GMM object, which contains the necessary parameters
+        needed to compute the Fisher vector.
+
+    References
+    -----------
+    .. [1] https://scikit-learn.org/stable/modules/generated/sklearn.mixture.GaussianMixture.html
+
+    Examples
+    --------
+    >>> from skimage.feature import fisher_vector
+    >>> sift_for_images = [np.random.random((10, 128)) for _ in range(10)]
+    >>> num_modes = 16
+    >>> # Estimate 16-mode GMM with these synthetic SIFT vectors
+    >>> gmm = fisher_vector.learn_gmm(sift_for_images, num_modes)
+    """
+
+    try:
+        from sklearn.mixture import GaussianMixture
+    except ImportError:
+        raise ImportError(
+            'scikit-learn is not installed. Please ensure it is installed in '
+            'order to use the Fisher vector functionality.'
+        )
+
+    if not isinstance(descriptors, (list, np.ndarray)):
+        raise DescriptorException(
+            'Please ensure descriptors are either a NumPY array, '
+            'or a list of NumPy arrays.'
+        )
+
+    d_mat_1 = descriptors[0]
+    if isinstance(descriptors, list) and not isinstance(d_mat_1, np.ndarray):
+        raise DescriptorException(
+            'Please ensure descriptors are a list of NumPy arrays.'
+        )
+
+    if isinstance(descriptors, list):
+        expected_shape = descriptors[0].shape
+        ranks = [len(e.shape) == len(expected_shape) for e in descriptors]
+        if (not all(ranks)):
+            raise DescriptorException(
+                'Please ensure all elements of your descriptor list '
+                'are of rank 2.'
+            )
+        dims = [e.shape[1] == descriptors[0].shape[1] for e in descriptors]
+        if not all(dims):
+            raise DescriptorException(
+                'Please ensure all descriptors are of the same dimensionality.'
+            )
+
+    if not isinstance(num_modes, int) or num_modes <= 0:
+        raise FisherVectorException(
+            'Please ensure num_modes is a positive integer.'
+        )
+
+    if 'covariance_type' in kwargs and kwargs['covariance_type'] != 'diag':
+        raise FisherVectorException('Covariane type must be "diag".')
+
+    if isinstance(descriptors, list):
+        descriptors = np.vstack(descriptors)
+
+    gmm = GaussianMixture(n_components=num_modes, covariance_type='diag')
+
+    gmm.fit(descriptors)
+
+    return gmm
+
+
+def compute_fisher_vector(descriptors, gmm, improved=False):
+    """Function to compute the Fisher vector given some descriptors/vectors,
+    and an associated estimated GMM.
+
+    Parameters
+    -----------
+    descriptors : np.ndarray
+        NumPy array of the descriptors to use to compute the Fisher vector
+        representation for.
+    gmm : sklearn.mixture.GaussianMixture
```

Comment:
Why is this not computed on behalf of the user?  I.e., why not have a single `fisher_vector` function call that does the whole thing?  Is it common, e.g., to compute the GMM and to re-use it several times?

Reply:
It is more common to compute is once and reuse. It's basically quantising all your training descriptors into GMM parameters. You will also need the GMM to compute the Fisher vectors for testing examples/images. It's the way vl-feat's Fisher vector implementation works as well (i.e. the separation of GMM from FV). See my reply [above](https://github.com/scikit-image/scikit-image/pull/5349#discussion_r626755166) for more detail.